product_main_supplierinfo/models/product_product.py

```diff
@@ -49,5 +49,7 @@ def _get_sellers(self):
             )
         )
         if not sellers:
-            sellers = all_sellers
+            sellers = all_sellers.filtered(lambda s: (s.product_id == self))
```

Comment:
this is already done above.
I would rewrite the whole method like this:
```
        sellers = all_sellers = self._prepare_sellers().filtered(
            lambda s: not s.company_id or s.company_id.id == self.env.company.id
        )
        # Filter by specific variant or none.
        by_variant = sellers.filtered(
            lambda s: s.product_id == self or not s.product_id
        )
        if by_variant:
            sellers = by_variant

        # Filter by dates
        today = fields.Date.context_today(self)
        by_date = sellers.filtered(lambda s: (
                s.date_start <= today if s.date_start else True
            ) and (
                s.date_end >= today if s.date_end else True
            )
        )
        if by_date:
            sellers = by_date
        
        # Fallback to all available sellers
        if not sellers:
            sellers = all_sellers
        return sellers
```

Reply:
also, we must have a test :wink:

Reply:
Seems in std Odoo they do not return a seller if the variant doesn't match the current one... https://github.com/odoo/odoo/blob/5e45d7267ddacee6cf882e1d3be60efc52b12967/addons/product/models/product.py#L621-L625
And only sellers sharing the same partner are selected, then sorted on the `price` to get the cheaper (`_prepare_sellers` is already doing some sort, don't know if we have to do it again...).

Reply:
I have added the test.
But I did not change the implementation.
About that line being already done above, not really because above the filtering is including the supplier info related to no variant. And the idea is if no valid (date wise) supplier info are found, then first return all supplier info only related to the variant.
This implementation change does not do that as the `by_variant` also includes the supplier info related to no variant.

Reply:
> Seems in std Odoo they do not return a seller if the variant doesn't match the current one... https://github.com/odoo/odoo/blob/5e45d7267ddacee6cf882e1d3be60efc52b12967/addons/product/models/product.py#L621-L625 And only sellers sharing the same partner are selected, then sorted on the `price` to get the cheaper (`_prepare_sellers` is already doing some sort, don't know if we have to do it again...).

It is `prepare_sellers` that is called above and not `select_sellers` which calls `prepare_sellers` in the odoo implementation.
Are you implying that we should change the implementation ?